design-pattern.md

```diff
@@ -429,3 +429,219 @@ public class Main {
+    // Crea una cassa curativa
+    Cassa cassaCurativa = 
+    new Cassa(() -> {
+      System.out.println("Sei stato curato!");
+    });
+    /* equivalente a:
+    Runnable effettoCassaCurativa = () -> {
+      System.out.println("Sei stato curato!");
+    };
+    Cassa cassaCurativa = 
+      new Cassa(cassaCurativa);
+    */
+
+    // Crea una cassa avvelenata
+    Cassa cassaAvvelenata = 
+    new Cassa(() -> {
+      System.out.println("Sei avvelenato!");
+    });
+
+    // Apri le casse curativa e avvelenata
+    cassaCurativa.apri();
+    cassaAvvelenata.apri();
+  }
+}
+```
+
+---
+
+### Observer
+
+**Observer** è un pattern che permette di notificare un oggetto quando un altro oggetto cambia.
+
+#### UML
+
+![](./immagini/observer.svg)
+
+In questo diagramma, troviamo `Subject` che è l'oggetto che cambia e notifica gli `Observer`; gli `Observer` che ricevono la notifica e si comportano conseguentemente. Infine abbiamo i `ConcreteObserver` che implementano `Observer` e definiscono il comportamento che si attiva quando `Subject` cambia. Inoltre `Subject` può essere astratta ed avere una `ConcreteSubject` che la estende.
+
+#### Esempio
+
+Supponiamo di voler costruire un sistema di gestione di dati di un'azienda. L'azienda ha un database che contiene i dati dei dipendenti, e un sistema che permette di visualizzare questi dati. Il sistema di visualizzazione deve essere aggiornato ogni volta che il database cambia. Usiamo un **Observer**.
+
+Per prima cosa, definiamo l'interfaccia `Observer`, in questo esempio la chiamiamo `Visualizzazione`, che definisce il metodo `aggiorna()`:
+
+```java
+public interface Visualizzazione {
+  public void aggiorna();
+}
+```
+
+Definiamo le varie classi `ConcreteObserver`, in questo esempio le chiamiamo `VisualizzazioneConsole`, `VisualizzazioneGUI`, ecc., che implementano `Visualizzazione`:
+
+```java
+public class VisualizzazioneConsole 
+implements Visualizzazione {
+  @Override
+  public void aggiorna() {
+    System.out.println(
+      "Aggiornamento della 
+      visualizzazione console");
+  }
+}
+```
+
+```java
+public class VisualizzazioneGUI 
+implements Visualizzazione {
+  @Override
+  public void aggiorna() {
+    System.out.println(
+      "Aggiornamento della 
+      visualizzazione GUI");
+  }
+}
+```
+
+Definiamo la classe `Subject`, in questo esempio la chiamiamo `Database`, che ha un `ArrayList` di `Visualizzazione` e un metodo `aggiornaVisualizzazioni()` che chiama il metodo `aggiorna()` di ogni `Visualizzazione`:
+
+```java
+public class Database {
+  private ArrayList<Visualizzazione> 
+    visualizzazioni;
+
+  public Database() {
+    visualizzazioni = new ArrayList<>();
+  }
+
+  public void aggiungiVisualizzazione
+  (Visualizzazione visualizzazione) {
+    visualizzazioni.add(visualizzazione);
+  }
+
+  public void rimuoviVisualizzazione
+  (Visualizzazione visualizzazione) {
+    visualizzazioni.remove(visualizzazione);
+  }
+
+  public void aggiornaVisualizzazioni() {
```

Comment:
Il metodo `aggiornaVisualizzazioni` dovrebbe essere `private` o ancora meglio `protected`, questo perché non deve essere il client a triggerare l'update di tutti gli Observer. Immagina in un caso d'uso reale, qui ci sarebbe tipo un metodo aggiungi dato (al DB dell'azienda) che al suo interno dovrebbe richiamare aggiornaVisualizzazioni. Se lo lasci pubblico, ogni volta che fa un update, il Client dovrebbe peroccuparsi di aggiornare tutti gli Observer (altrimenti non avrebbe senso lasciarlo pubblico se viene già richiamato all'interno della classe) o potrebbe richiamare update quando non è necessario.
Se ti va aggiungi un metodo aggiungiDato che richiama aggiornaVisualizzazioni e cambia il main di conseguenza, altrimenti se non hai voglia o per risparmiare tempo puoi giusto mettere `aggiornaVisualizzazioni` protected e cancellare la parte con il `main`, tanto credo che l'Observer sia abbastanza comprensibile.
P.S. : Non mi odiare perfavore